File: swift-sdk/Sources/GistAdsSDK/Models/DisplayAdLoadState.swift
//
//  DisplayAdLoadState.swift
//  GistAdsSDK
//
//  Pure state-derivation logic for GistDisplayAdControl, factored out of the
//  SwiftUI view so it can be unit tested without needing to render a view
//  hierarchy.
//

import Foundation

/// The possible states `GistDisplayAdControl` can be in while loading a display ad.
enum DisplayAdLoadState: Equatable {
    case loading
    case loaded(String)
    case noFill
    case failed(String)

    /// Derive the resulting state from a `DisplayAdAPIService.fetchAd` outcome.
    /// - Parameter result: The result of attempting to fetch and render an ad.
    /// - Returns: `.loaded` on success, `.noFill` for a no-fill response, or
    ///   `.failed` for any other error.
    static func from(result: Result<String, Error>) -> DisplayAdLoadState {
        switch result {
        case .success(let content):
            return .loaded(content)
        case .failure(let error):
            if case DisplayAdAPIError.noFill = error {
                return .noFill
            }
            return .failed(error.localizedDescription)
        }
    }
}

File: android-sdk/src/main/java/com/gist/ads/sdk/models/DisplayAdLoadState.kt
package com.gist.ads.sdk.models

import com.gist.ads.sdk.services.DisplayAdAPIException

/**
 * Pure state-derivation logic for `GistDisplayAdControl`, factored out of the
 * composable so it can be unit tested without needing to render a UI.
 */
sealed class DisplayAdLoadState {
    object Loading : DisplayAdLoadState()
    data class Loaded(val html: String) : DisplayAdLoadState()
    object NoFill : DisplayAdLoadState()
    data class Failed(val message: String) : DisplayAdLoadState()

    companion object {
        /**
         * Derive the resulting state from a `DisplayAdAPIService.fetchAd` outcome.
         * @return [Loaded] on success, [NoFill] for a no-fill response, or
         *   [Failed] for any other error.
         */
        fun from(result: Result<String>): DisplayAdLoadState {
            return result.fold(
                onSuccess = { content -> Loaded(content) },
                onFailure = { error ->
                    if (error is DisplayAdAPIException.NoFill) {
                        NoFill
                    } else {
                        Failed(error.message ?: "Unknown error")
                    }
                }
            )
        }
    }
}
